Avkr003/leverage improvements

avkr003 opened the pull request:
- [Adding validation in params](https://github.com/elys-network/elys/commit/c7e82a2ee54c3a9024eabbeb83688ba2484e30eb)
- [fixing data type for msgStopLoss](https://github.com/elys-network/elys/commit/b5462afb433834e62072d6ce53cda19db5b5c31a)
- [removing redundant code, improving test cases for msgs and param, fixing nomenclature](https://github.com/elys-network/elys/commit/3b96414adc487b67dc813ab2733b3710d3f3c5fa)
- [fixing add/remove pool tx logic, improving validate basic of msgs and add all their test cases](https://github.com/elys-network/elys/commit/ff8e777e26a0dacf726b71bad9f391805114fe2e)
- [improving positions validate functions](https://github.com/elys-network/elys/commit/862bf1f897a74a17e589ee290c7a483b15d7521f)
- [optimizing BeginBlocker for leveragelp, improving it's nomenclature, increasing it's test coverage to 90%](https://github.com/elys-network/elys/commit/08b6dfcbd42056c64e707974bb67a72208a7eefb)
- [adding test cases for close positions tx](https://github.com/elys-network/elys/commit/9dd346f36fb8fa10db481978acd2a178027f3f53)
- [[critical] Removed IsPoolEnabled function which add pool without governance](https://github.com/elys-network/elys/commit/814fd80f4f610233f6e206717a5ae307ef0601bf)
- [adding more validations](https://github.com/elys-network/elys/commit/a7fe542c8d5a87e6c6cd3242097a686ae671f179)


Key Changes:

- Earlier when adding a pool to leverage LP, we were not validating with module params max allowed and when opening position, we were directly comparing to module params from user tx. Now, we set pool leverage max by comparing with module params and while opening position, we compare with pool leverage max, not module params max. [Add Pool](https://github.com/elys-network/elys/pull/752/files#diff-7440c0115abcfd190a1b1d48a7225b4c1e124f8714097091fbce23d15cb186a6R32) [Open position](https://github.com/elys-network/elys/pull/752/files#diff-81b38e59862cb65103df4a45e2538ba735dc41171960f88e94fc3fafdd61af63R21)

- CheckAndLiquidateUnhealthyPosition and CheckAndCloseAtStopLoss always return an error if the position is not closed. Changed return variable `earlyReturn` to `closeAttempted` and changed logic everywhere else accordingly. `earlyReturn`  was ambiguous where it was returning from.  `closeAttempted` returns true if `ForceCloseLong` has been called in the function.

- Remove `IsPoolEnabled` and `IsPoolClosed` functions as it was adding a pool without any governance proposal 

- Added msg basic validations
   - In msg_close_positions, not allowing repeating close position IDs including both stop loss and unhealthy positions
   - 0 stop loss price while opening position is allowed indicating not set by the user. (based on testnet app) [here](https://github.com/elys-network/elys/pull/752/files#diff-f70cc5f2a83899b1ac1e3c8ced6279691243512135652e3df34a4d4e19165a63R124)
   
 
   - Not allow same position ids while claiming rewards [here](https://github.com/elys-network/elys/pull/752/files#diff-f70cc5f2a83899b1ac1e3c8ced6279691243512135652e3df34a4d4e19165a63R385)
   
   - Params validation:
     - Not allowing negative leverage max
     - PoolOpenTheshold has to be always +ve (0 not allowed)

amityadav0 commented on x/leveragelp/keeper/begin_blocker.go:
```diff
@@ -67,15 +80,15 @@ func (k Keeper) LiquidatePositionIfUnhealthy(ctx sdk.Context, position *types.Po
 	if err != nil {
 		ctx.Logger().Error(errors.Wrap(err, fmt.Sprintf("error updating position health: %s", position.String())).Error())
-		return false, true, sdk.ZeroDec(), err
+		return false, false, sdk.ZeroDec(), err
```
these conditions are used in `closePositons`, could you please verify changing this to false doesn't impact that part ?

avkr003 replied:
No, it doesn't. Instead of returning the variable name `earlyReturn`, I changed it to `closeAttempted`. 
`earlyReturn` was ambiguous - it wasn't very clear from where it's returning.

amityadav0 commented on x/leveragelp/keeper/begin_blocker.go:
```diff
@@ -92,10 +105,10 @@ func (k Keeper) LiquidatePositionIfUnhealthy(ctx sdk.Context, position *types.Po
 		sdk.NewAttribute("health", position.PositionHealth.String()),
 	))
-	return isHealthy, false, h, nil
+	return isHealthy, true, h, nil
```
same

avkr003 replied:
same as above

amityadav0 commented on x/leveragelp/keeper/begin_blocker.go:
```diff
@@ -106,19 +119,19 @@ func (k Keeper) ClosePositionIfUnderStopLossPrice(ctx sdk.Context, position *typ
 	lpTokenPrice, err := ammPool.LpTokenPrice(ctx, k.oracleKeeper)
 	if err != nil {
-		return false, true, err
+		return false, false, err
```
used in close_positions

avkr003 replied:
Checking there through error. If this function doesn't close the position, it always return an error

amityadav0 commented on x/leveragelp/keeper/msg_server_close_positions.go:
```diff
@@ -12,46 +13,69 @@ import (
 	ctx := sdk.UnwrapSDKContext(goCtx)
 
+	poolMap := make(map[uint64]types.Pool)
+	ammPoolMap := make(map[uint64]ammtypes.Pool)
```
this optmisation may not work, added my thoughts above

avkr003 commented:
Key Changes:

- Earlier when adding a pool to leverage LP, we were not validating with module params max allowed and when opening position, we were directly comparing to module params from user tx. Now, we set pool leverage max by comparing with module params and while opening position, we compare with pool leverage max, not module params max. [Add Pool](https://github.com/elys-network/elys/pull/752/files#diff-7440c0115abcfd190a1b1d48a7225b4c1e124f8714097091fbce23d15cb186a6R32) [Open position](https://github.com/elys-network/elys/pull/752/files#diff-81b38e59862cb65103df4a45e2538ba735dc41171960f88e94fc3fafdd61af63R21)

- CheckAndLiquidateUnhealthyPosition and CheckAndCloseAtStopLoss always return an error if the position is not closed. Changed return variable `earlyReturn` to `closeAttempted` and changed logic everywhere else accordingly. `earlyReturn`  was ambiguous where it was returning from.  `closeAttempted` returns true if `ForceCloseLong` has been called in the function.

- Remove `IsPoolEnabled` and `IsPoolClosed` functions as it was adding a pool without any governance proposal 

- Added msg basic validations
   - In msg_close_positions, not allowing repeating close position IDs including both stop loss and unhealthy positions
   - 0 stop loss price while opening position is allowed indicating not set by the user. (based on testnet app) [here](https://github.com/elys-network/elys/pull/752/files#diff-f70cc5f2a83899b1ac1e3c8ced6279691243512135652e3df34a4d4e19165a63R124)
   
 
   - Not allow same position ids while claiming rewards [here](https://github.com/elys-network/elys/pull/752/files#diff-f70cc5f2a83899b1ac1e3c8ced6279691243512135652e3df34a4d4e19165a63R385)
   
   - Params validation:
     - Not allowing negative leverage max
     - PoolOpenTheshold has to be always +ve (0 not allowed)
